test/test_trio_server/test_full_sync.py

```diff
@@ -0,0 +1,47 @@
+import unittest
+from test.test_trio_server.mock_lean import \
+    LeanShouldGetRequest, LeanShouldNotGetRequest, LeanSendsResponse, start_with_mock_lean
+from lean_client.trio_server import TrioLeanServer
+import trio
+import trio.testing
+
+
+class TestFullSync(unittest.TestCase):
+    def test_full_sync_waits_until_lean_ready(self):
+        """
+        Check that full_sync waits until the Lean server is ready.
+        In particular, it must first get a "file invalidated" response, and then an "all_tasks" response
+        before continuing.
+        """
+        mock_lean_script = [
+            LeanShouldGetRequest({"file_name": "test.lean", "content": '--', "seq_num": 1, "command": "sync"}),
```

Comment:
I think you'll be tired of writing all this Json. It would make more sense to either pass a subclass of `Request` object or have arguments to the `LeanShould...` constructor.

Reply:
In practice I usually just copy the json from the communication of a real world example.  

I also think this is a little dangerous though to always pass in the higher level object instead of the raw json.  Consider the following case:  We start working with, say, hole requests and discover that we forgot or misspelled a field in the hole request.  A test like this would ensure that exactly the right fields are sent in the correct way, instead of copying any errors we included in `.to_json()`.   Of course, this is a special case.  Usually, I'm just checking that Mock Lean and the interface code both get to the same point in their communication.  Maybe it is worth having two versions of LeanShould.  One which takes the JSON when it is needed, and another which takes the Request object.  

As for LeanSendsResponse, I think it is more important that we keep it as JSON at least for some of the cases.  For example, in later PRs I am testing that the interface doesn't crash when being send an infosource without a line or column.  Again, however, we could have multiple versions of LeanSendsRequest.  

What do you think?

Reply:
I think you are mixing tests here. There should be unit tests checking that `comands.py` does its job correctly, and then the tests we are currently looking at should trust it.

Reply:
Ok, I'll make an issue that we need to add unit tests for `commands.py`.  I guess there is no real reason to work on that until we complete https://github.com/leanprover-community/lean-client-python/issues/6.

Reply:
Ok, I added issue #11

Reply:
Ok, so I tried to refactor this so that `LeanShouldGetRequest` and `LeanShouldGetRequest` take objects from the `commands.py`.  It works for `LeanShouldGetRequest`.  Previously, the command was:
```python
LeanShouldGetRequest({"file_name": "test.lean", "content": '--', "seq_num": 1, "command": "sync"}),
```
and after it is
```python
LeanShouldGetRequest(SyncRequest(file_name="test.lean", content="--"), seq_num=1)
```
I'm not sure it is saving a lot of typing, but it does add some type safety.

As for `LeanShouldGetResponse`, that is more complicated.  I would need to write code that turns each `Response` object into a dictionary.  Maybe this can be revisited after #6 , but for now I'd like to just finish this up so I can fix the other bugs in the server.